### DISCUSSION TITLE
Is `--parser` necessary?

### DISCUSSION THREAD

**Discussion by ryuheechul** [TomWright/dasel#154] [Q&A] [answered]
Yesterday (well I was introduced to `dasel` yesterday!), I was confused with `-p` option and you can read about what mistake I made in here, https://github.com/TomWright/dasel/issues/153.

Now I know `-p` is shorthand for `--parser`, thanks to the answer in the issue above 😄 .

However I think it's possible to get rid of `-p` all together if we use `-r` (`--read`) value to result as a default value to print what output format to print. (Which I think what most people would expect to happen)

For example,
Currently `echo "who: i'm yaml" | dasel -p yaml` should just work with  `echo "who: i'm yaml" | dasel -r yaml`.
And in fact, it already does that now. (same with `echo '{"who": "i am json"}' | dasel -r json`)

When you need to change the output to `json`, you will need to `echo "who: i'm yaml" | dasel -r yaml -w json` anyway.
So I don't see `-p` is doing what other options can't in both situations above.

Thus, to the original point. Why do we need `--parser`?

**Comment by TomWright** [comment, marked as answer]
The ability to convert between data types was added in `v1.7.0`: https://github.com/TomWright/dasel/blob/master/CHANGELOG.md#v170---2020-11-30

Prior to this the `-p` flag was used to tell dasel which data type to expect when reading from `stdin`.

In order to keep backward compatibility the flag has to remain but as you say, you can essentially forget about it and just use `-r` since internally when no `-w` flag is provided, the `-r` value is used.

- `-r` is shorthand for `--read`, which is used to set a read parser.
- `-w` is shorthand for `--write`, which is used to set a write parser.
- `-p` is shorthand for `--parser`, which is effectively an alias of `-r <val> -w <val>`
